File: CSharp/GameOfLife/GridManager.cs
namespace GameOfLife
{
    public class GridManager
    {
        public Cell[,] GetNewState(Cell[,] grid)
        {
            var newGrid = new Cell[grid.GetLength(0), grid.GetLength(1)];
            for (int i = 0; i < grid.GetLength(0); i++)
            {
                for (int j = 0; j < grid.GetLength(1); j++)
                {
                    newGrid[i, j] = new Cell(false);
                }
            }

            for (int i = 0; i < grid.GetLength(0); i++)
            {
                for (int j = 0; j < grid.GetLength(1); j++)
                {
                    int nbNeighbours = CountNeighbours(grid, i, j);
                    if (nbNeighbours == 2)
                    {
                        newGrid[i, j].IsAlive = grid[i, j].IsAlive;
                    }
                    else if (nbNeighbours == 3)
                    {
                        newGrid[i, j].IsAlive = true;
                    }
                }
            }
            return newGrid;
        }

        public int CountNeighbours(Cell[,] grid, int i, int j)
        {
            int nbNeighbours = 0;
            if (IsAliveNeighbour(grid, i - 1, j - 1))
            {
                nbNeighbours++;
            }
            if (IsAliveNeighbour(grid, i, j - 1))
            {
                nbNeighbours++;
            }
            if (IsAliveNeighbour(grid, i + 1, j - 1))
            {
                nbNeighbours++;
            }
            if (IsAliveNeighbour(grid, i - 1, j))
            {
                nbNeighbours++;
            }
            if (IsAliveNeighbour(grid, i + 1, j))
            {
                nbNeighbours++;
            }
            if (IsAliveNeighbour(grid, i - 1, j + 1))
            {
                nbNeighbours++;
            }
            if (IsAliveNeighbour(grid, i, j + 1))
            {
                nbNeighbours++;
            }
            if (IsAliveNeighbour(grid, i + 1, j + 1))
            {
                nbNeighbours++;
            }

            return nbNeighbours;
        }

        private static bool IsAliveNeighbour(Cell[,] grid, int i, int j)
        {
            return j >= 0 &&
                j < grid.GetLength(1) &&
                i >= 0 &&
                i < grid.GetLength(0) &&
                grid[i, j].IsAlive;
        }
    }
}

File: scala/src/main/scala/gameOfLife/GridManager.scala
package gameOfLife

class GridManager {

  def GetNewState(grid: Array[Array[Cell]]): Array[Array[Cell]] = {
    var newGrid = Array.ofDim[Cell](grid.length, grid(0).length)
    for (i <- 0 until grid.length) {
      for (j <- 0 until grid(0).length) {
        newGrid(i)(j) = new Cell(false);
      }
    }

    for (i <- 0 until grid.length) {
      for (j <- 0 until grid(0).length) {
        var nbNeighbours: Int = CountNeighbours(grid, i, j);
        if (nbNeighbours == 2) {
          newGrid(i)(j).isAlive = grid(i)(j).isAlive;
        }
        else if (nbNeighbours == 3) {
          newGrid(i)(j).isAlive = true;
        }
      }
    }
    return newGrid;
  }

  def CountNeighbours(grid: Array[Array[Cell]], i: Int, j: Int): Int = {
    import GridManager._

    var nbNeighbours: Int = 0;
    if (isAliveNeighbour(grid, i - 1, j - 1)) {
      nbNeighbours += 1;
    }
    if (isAliveNeighbour(grid, i, j - 1)) {
      nbNeighbours += 1;
    }
    if (isAliveNeighbour(grid, i + 1, j - 1)) {
      nbNeighbours += 1;
    }
    if (isAliveNeighbour(grid, i - 1, j)) {
      nbNeighbours += 1;
    }
    if (isAliveNeighbour(grid, i + 1, j)) {
      nbNeighbours += 1;
    }
    if (isAliveNeighbour(grid, i - 1, j + 1)) {
      nbNeighbours += 1;
    }
    if (isAliveNeighbour(grid, i, j + 1)) {
      nbNeighbours += 1;
    }
    if (isAliveNeighbour(grid, i + 1, j + 1)) {
      nbNeighbours += 1;
    }

    return nbNeighbours;
  }

}

object GridManager {
  def isAliveNeighbour(grid: Array[Array[Cell]], i: Int, j: Int): Boolean = {
    return j >= 0 &&
      j < grid(0).length &&
      i >= 0 &&
      i < grid.length &&
      grid(i)(j).isAlive;
  }
}
